src/data/set/prod.lean

```diff
@@ -315,6 +315,9 @@ lemma mem_diagonal (x : α) : (x, x) ∈ diagonal α := by simp [diagonal]
 
 @[simp] lemma mem_diagonal_iff {x : α × α} : x ∈ diagonal α ↔ x.1 = x.2 := iff.rfl
 
+instance decidable_mem_diagonal [h : decidable_eq α] : decidable_pred (diagonal α) :=
```

Comment:
Right, this seems OK too - my main concern was exactly as you said (just because a set is def-eq to predicates doesn't mean you should necessarily treat it like that...)